### PR TITLE
Updated starters, with doc and CB3 updated to consume them

### PR DESCRIPTION
**Pull request by markxnelson**
This PR adds the 23.4.0 starters (for SB3.0+), updates CB3 (SPB3) to use them, and adds documentation for starters.

### CODE REVIEW

**Review by pasimoes — APPROVED** [review]

**Review by andytael — DISMISSED** [review]

File: cloudbank-v3/spring-apps-spring3/checks/src/main/resources/application.yaml [hunk 1]
```diff
@@ -4,8 +4,7 @@ spring:
   zipkin:
     base-url: ${zipkin.base-url}
 
-oracle:
-  aq:
+  datasource:
```
Comment (rhondaday) [inline cloudbank-v3/spring-apps-spring3/checks/src/main/resources/application.yaml:7]: datasource -> data source

File: docs-source/spring/content/starters/_index.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+---
+title: "Oracle Spring Boot Starters"
+---
+
+Oracle provides a number of Spring Boot Starters that make it easy to use various Oracle technologies in Spring Boot projects.
+
+The Oracle Spring Boot Starters are available in [Maven Central](https://central.sonatype.com/namespace/com.oracle.database.spring) and include
+the following starters.
+
+**Note**: The versioning of starters was originally in line with the matching Spring Boot version, but starting in Noveember, 2023, the version
```
Comment (rhondaday) [inline docs-source/spring/content/starters/_index.md:10]: typo: Noveember

File: docs-source/spring/content/starters/_index.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+numbering scheme was changed to more closely align with Oracle Database version numbering.  The 23.4.0 starter is supported for Spring Boot 3.0
```
Comment (rhondaday) [inline docs-source/spring/content/starters/_index.md:11]: One space between sentences.

File: docs-source/spring/content/starters/aqjms.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
+---
+title: "Oracle Spring Boot Starter for AQ/JMS"
+---
+
+This starter provides support for Oracle Transactional Event Queues and Oracle Advanced Queues
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:5]: Use:
Oracle Transactional Event Queues (TEQ)
Oracle Advanced Queuing (AQ)

File: docs-source/spring/content/starters/aqjms.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+as JMS providers.  It depends on the UCP starter.
+
+**Note**: By default, the Data Source and JMS Connection Factory that the starter injects into
+your application will share the same database transaction.  This means that you can start a
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:9]: Change "will share" to "shares".

File: docs-source/spring/content/starters/aqjms.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+transaction, read from a queue, perform an update operation, and then commit or rollback that
+whole unit of work, including the message consumption, for example.
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:11]: Delete ", for example"

File: docs-source/spring/content/starters/aqjms.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+To add this starter to your project, add this Maven dependency:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:13]: "... add this Mavin dependency. For example:"

File: docs-source/spring/content/starters/aqjms.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+```xml
+<dependency>
+    <groupId>com.oracle.database.spring</groupId>
+    <artifactId>oracle-spring-boot-starter-aqjms</artifactId>
+    <version>23.4.0</version>
+    <type>pom</type>
+</dependency>
+```
+
+For Gradle projects, add this dependency:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:24]: "... add this dependency. For example:"

File: docs-source/spring/content/starters/aqjms.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+```
+implementation 'com.oracle.database.spring:oracle-spring-boot-starter-aqjms:23.4.0:pom'
+```
+
+To configure your application to use Oracle Transacational Event Queues or Oracle Advanced
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:30]: typo: Transacational -> Transactional

File: docs-source/spring/content/starters/aqjms.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+Queuing, you must annotate you application with the `@EnableJms` annotation, and create the
+two following beans:
+
+* A `JmsListenerContainerFactory<?>` bean, which can be created as shown below.  Note that
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:34]: "... as shown in the following example."

File: docs-source/spring/content/starters/aqjms.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+  you can override settings if you need to.  Also, note that the name of the method defines
+  the name of the factory, which you will when creating JMS listeners.
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:36]: "... which you will [what] ...?

File: docs-source/spring/content/starters/aqjms.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+	  ConfigurableApplicationContext context = SpringApplication.run(JmsSampleApplication.class, args);
+	}
+  
+}
+```
+
+To send a message to a JMS queue or topic, get an instance of the `JmsTemplate` from the Spring
+Application context, and call the `convertAndSend()` method specifying the name of the queue or
+topic, and providing the object to be converted and sent in the payload of the message, as shown
+in the example below:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:94]: "... in the following example:"

File: docs-source/spring/content/starters/aqjms.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+```java
+JmsTemplate jmsTemplate = context.getBean(JmsTemplate.class);
+jmsTemplate.convertAndSend("mailbox", new Email(-1, "info@example.com", "Hello"));
+```
+
+To receive messages from a JMS queue or topic, create a method that takes your message class, e.g., `Email`,
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:101]: e.g. -> for example

File: docs-source/spring/content/starters/aqjms.md [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+as input.  Annotate the method with the `@JmsListener` annotation, specifying the destination, i.e., the name of
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:102]: i.e. -> that is

File: docs-source/spring/content/starters/aqjms.md [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+the queue or topic, and the container factory name that you created earlier, as shown in the example below:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:103]: "... in the following example:"

File: docs-source/spring/content/starters/aqjms.md [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+    @JmsListener(destination = "mailbox", containerFactory = "myFactory")
+    public void receiveMessage(Email email) {
+      System.out.println("Received <" + email + ">");
+    }
+
+}
+```
+
+Note that the starter will use the configuration for `spring.datasource` as the connection
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:122]: will use -> uses

File: docs-source/spring/content/starters/aqjms.md [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+details for the Oracle Database hosting the queues and topics.  If you wish to use a different
+one, you must use a named configuration, e.g., `spring.datasource.txeventq` and use Java
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:124]: one -> configuration
e.g. -> for example

File: docs-source/spring/content/starters/aqjms.md [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+configuration (as shown above for the UCP starter) and annotate the configuration with
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:125]: "(as previously shown ..."

File: docs-source/spring/content/starters/aqjms.md [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+the standard Spring `@Qualifier` annotation, specifying the correct name, e.g., `txevevntq`.
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:126]: e.g. -> for example

File: docs-source/spring/content/starters/aqjms.md [hunk 20]
```diff
@@ -0,0 +1,127 @@
+---
+title: "Oracle Spring Boot Starter for AQ/JMS"
+---
+
+This starter provides support for Oracle Transactional Event Queues and Oracle Advanced Queues
+as JMS providers.  It depends on the UCP starter.
+
+**Note**: By default, the Data Source and JMS Connection Factory that the starter injects into
```
Comment (rhondaday) [inline docs-source/spring/content/starters/aqjms.md:8]: Data Source -> data source

File: docs-source/spring/content/starters/ucp.md [hunk 21]
```diff
@@ -0,0 +1,104 @@
+---
+title: "Oracle Spring Boot Starter for Universal Connection Pool"
+---
+
+This starter provides a connection (data source) to an Oracle Database using Universal Connection Pool, which provides an efficient way
+to use database connections. 
+
+To add this starter to your project, add this Maven dependency:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:8]: "... dependency. For example:"

File: docs-source/spring/content/starters/ucp.md [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+```xml
+<dependency>
+    <groupId>com.oracle.database.spring</groupId>
+    <artifactId>oracle-spring-boot-starter-ucp</artifactId>
+    <version>23.4.0</version>
+    <type>pom</type>
+</dependency>
+```
+
+For Gradle projects, add this dependency:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:19]: "... dependency. For example:"

File: docs-source/spring/content/starters/ucp.md [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+
+```
+implementation 'com.oracle.database.spring:oracle-spring-boot-starter-ucp:23.4.0:pom'
+```
+
+An Oracle data source will be injected into your application and can be used normally. You must configure
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:25]: will be -> is

File: docs-source/spring/content/starters/ucp.md [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+it as shown below, and you should also add either Spring Data JDBC or Spring Data JPA to your project.
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:26]: What is "it" referring to? The data source? The application?

"... as shown in the following example, and ..."

File: docs-source/spring/content/starters/ucp.md [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+
+To configure the datasource, provide a `spring.datasource` object in your Spring `application.yaml`,
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:28]: datasource -> data source

File: docs-source/spring/content/starters/ucp.md [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+or equivalent, as shown in the example below.  The `oracleucp` entry is optional, and can be used
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:29]: "... in the following example."

File: docs-source/spring/content/starters/ucp.md [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+to fine tune the configuration of the connection pool if desired.  For details of available settings,
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:30]: "... pool, if desired."

File: docs-source/spring/content/starters/ucp.md [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+please refer to the [JavaDoc](https://docs.oracle.com/en/database/oracle/oracle-database/21/jjuar/oracle/ucp/jdbc/UCPDataSource.html).
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:31]: Delete "please"

File: docs-source/spring/content/starters/ucp.md [hunk 29]
```diff
@@ -0,0 +1,104 @@
...
+    type: oracle.ucp.jdbc.PoolDataSource
+    oracleucp:
+      connection-factory-class-name: oracle.jdbc.pool.OracleDataSource
+      connection-pool-name: AccountConnectionPool
+      initial-pool-size: 15
+      min-pool-size: 10
+      max-pool-size: 30
+```
+
+The `spring.datasource.url` can be in the basic format (as shown above), or in TNS format if your application
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:57]: "... format (as previously shown), or ..."

Use:
Transparent Network Substrate (TNS)

after which you can use TNS

File: docs-source/spring/content/starters/ucp.md [hunk 30]
```diff
@@ -0,0 +1,104 @@
...
+uses TNS. 
+
+Note that the connections to the database will use the `DEDICATED` server by default.  If you wish to use `SHARED`
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:60]: Delete "will"

File: docs-source/spring/content/starters/ucp.md [hunk 31]
```diff
@@ -0,0 +1,104 @@
...
+or `POOLED` you can append that to the basic URL, or add it to the TNS names entry.  For example, to used
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:61]: Add a comma before "you"

File: docs-source/spring/content/starters/ucp.md [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+      min-pool-size: 10
+      max-pool-size: 30
+```
+
+The `spring.datasource.url` can be in the basic format (as shown above), or in TNS format if your application
+uses TNS. 
+
+Note that the connections to the database will use the `DEDICATED` server by default.  If you wish to use `SHARED`
+or `POOLED` you can append that to the basic URL, or add it to the TNS names entry.  For example, to used 
+database resident pooled connections, you would change the URL shown in the example above to this:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:62]: "... in the previous example to the following:"

File: docs-source/spring/content/starters/ucp.md [hunk 33]
```diff
@@ -0,0 +1,104 @@
...
+Note that the connections to the database will use the `DEDICATED` server by default.  If you wish to use `SHARED`
+or `POOLED` you can append that to the basic URL, or add it to the TNS names entry.  For example, to used 
+database resident pooled connections, you would change the URL shown in the example above to this: 
+
+```yaml
+  datasource:
+    url: jdbc:oracle:thin:@//myhost:1521/pdb1:pooled
+```
+
+If you are using TNS, add `server=pooled` to the `connect_data` as shown below:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:69]: "If ... to the 'connect_data'. For example:"

File: docs-source/spring/content/starters/ucp.md [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+```
+mydb_tp = (description=
+ (retry_count=20)
+ (retry_delay=3)
+ (address=(protocol=tcps)(port=1521)(host=myhost))
+ (connect_data=(service_name=pdb1)(server=pooled))
+ (security=(ssl_server_dn_match=yes)))
+```
+
+If you prefer to use Java configuration, the data source can be configured as shown in this example:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/ucp.md:80]: "... in the following example:"

File: docs-source/spring/content/starters/wallet.md [hunk 35]
```diff
@@ -0,0 +1,36 @@
+---
+title: "Oracle Spring Boot Starter for Wallet"
+---
+
+This starter provides support for wallet-based authentication for Oracle Database connections.  It depends
+on the UCP starter.
+
+To add this starter to your project, add this Maven dependency:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/wallet.md:8]: "... add this Maven dependency. For example:"

File: docs-source/spring/content/starters/wallet.md [hunk 36]
```diff
@@ -0,0 +1,36 @@
...
+```xml
+<dependency>
+    <groupId>com.oracle.database.spring</groupId>
+    <artifactId>oracle-spring-boot-starter-wallet</artifactId>
+    <version>23.4.0</version>
+    <type>pom</type>
+</dependency>
+```
+
+For Gradle projects, add this dependency:
```
Comment (rhondaday) [inline docs-source/spring/content/starters/wallet.md:19]: "... dependency. For example:"

File: docs-source/spring/content/starters/wallet.md [hunk 37]
```diff
@@ -0,0 +1,36 @@
...
+
+```
+implementation 'com.oracle.database.spring:oracle-spring-boot-starter-wallet:23.4.0:pom'
+```
+
+You need to provide the wallet to your application.  You can specify the location in the `spring.datasource.url`
+as shown below.
```
Comment (rhondaday) [inline docs-source/spring/content/starters/wallet.md:26]: "... as shown in the following example:"

File: docs-source/spring/content/starters/wallet.md [hunk 38]
```diff
@@ -0,0 +1,36 @@
...
+
+```
+jdbc:oracle:thin:@mydb_tp?TNS_ADMIN=/oracle/tnsadmin
+```
+
+Note that the location specified in the `sqlnet.ora` must match the actual location of the file.
+
+If your service is deployed in Kubernetes, the wallet should be placed in a Kubernetes secret which
```
Comment (rhondaday) [inline docs-source/spring/content/starters/wallet.md:34]: "If ... Kubernetes, then the ..."

**Review by rhondaday — COMMENTED** [review]
A few suggestions.

**Review by pasimoes — APPROVED** [review]

**Pull request merged**